engineering-plans/0001-sql-query-combination.org

```diff
@@ -0,0 +1,397 @@
+#+end_src
+
+** Handling interfaces
+
+If `parents or `children` (or both) are interfaces, we resolve the
+interfaces into the concrete types implementing them, produce a query for
+each combination of parent/child concrete type and combine those queries
+via `union all`.
+
+Since implementations of the same interface will generally differ in the schema
+they use, we can not form a `union all` of all the data in the tables for
+these concrete types, but have to first query only attributes that we know
+will be common to all entities implementing the interface, most notably the
+`vid` (a unique identifier that identifies the precise version of an
+entity), and then later fill in the details of each entity by converting it
+directly to JSON.
+
+That means that when we deal with children that are an interface, we will
+first select only the following columns (where exactly they come from
+depends on how the parent/child relationship is modeled)
+#+begin_src sql
+  select '{__typename}' as entity, c.vid, c.id, parent_id
+#+end_src
+
+and form the `union all` of these queries. We then use that union to rank
+children as described above.
+
+** Handling parent/child relationships
+
+How we get the children for a set of parents depends on how the
+relationship between the two is modeled. The interesting parameters there
+are whether parents store a list or a single child, and whether that field
+is derived, together with the same for children.
+
+There are a total of 16 combinations of these four boolean variables; four
+of them, when both parent and child derive their fields, are not
+permissible. It also doesn't matter whether the child derives its parent
+field: when the parent field is not derived, we need to use that since that
+is the only place that contains the parent -> child relationship. When the
+parent field is derived, the child field can not be a derived field.
+
+That leaves us with the following combinations of whether the parent and
+child store a list or a scalar value, and whether the parent is derived:
+
+| Case | Parent list? | Parent derived? | Child list? | Join cond                  | Type |
+|------+--------------+-----------------+-------------+----------------------------+------|
+|    1 | TRUE         | TRUE            | TRUE        | child.parents ∋ parent.id  | A    |
+|    2 | FALSE        | TRUE            | TRUE        | child.parents ∋ parent.id  | A    |
+|    3 | TRUE         | TRUE            | FALSE       | child.parent = parent.id   | B    |
+|    4 | FALSE        | TRUE            | FALSE       | child.parent = parent.id   | B    |
+|    5 | TRUE         | FALSE           | TRUE        | child.id ∈ parent.children | C    |
+|    6 | TRUE         | FALSE           | FALSE       | child.id ∈ parent.children | C    |
+|    7 | FALSE        | FALSE           | TRUE        | child.id = parent.child    | D    |
+|    8 | FALSE        | FALSE           | FALSE       | child.id = parent.child    | D    |
+
```

Comment:
For formatting this, we can use https://github.github.com/gfm/#tables-extension- in mdBook.